zjit/src/codegen.rs

```diff
@@ -1073,8 +1073,17 @@ fn gen_guard_type(jit: &mut JITState, asm: &mut Assembler, val: lir::Opnd, guard
     } else if let Some(expected_class) = guard_type.runtime_exact_ruby_class() {
         asm_comment!(asm, "guard exact class");
 
-        // Get the class of the value
-        let klass = asm.ccall(rb_yarv_class_of as *const u8, vec![val]);
+        // Check if it's a special constant
+        asm.test(val, (RUBY_IMMEDIATE_MASK as u64).into());
+        asm.jnz(side_exit(jit, state, GuardType(guard_type))?);
+
+        // Check if it's false
+        assert!(Qfalse.as_i64() == 0);
+        asm.test(val, val);
+        asm.jz(side_exit(jit, state, GuardType(guard_type))?);
+
+        // Load the class from the object's klass field
+        let klass = asm.load(Opnd::mem(64, val, RUBY_OFFSET_RBASIC_KLASS));
 
         asm.cmp(klass, Opnd::Value(expected_class));
         asm.jne(side_exit(jit, state, GuardType(guard_type))?);
```

Comment:
I'd use a local for the side exit target to show that all these jump to same side exit.